README.md
## Angular BarChart Directive

A reusable angular.js directive that will render a bar chart based upon data passed from the controller.

In your app config, add the directive as a dependancy injection like so

```javascript
app = angular.module('appName', [
   'ftBarChart'
]);
```

Use the directive by placing the code tag as follows:

```html
<div bar-chart ng-model="data"></div>
```

Defaults are in place for all the chart configurations, however the following configurations may be added as attributes to the directive element.

```javascript
chart-height="200" 
chart-width="778" 
chart-color="rgba(99,99,99,0.3)"
```

The data that fills the chart is passed by reference to the directive from the scope of the parent controller defined by the ng-model attribute.

```javascript
// Controller File
app.controller('someCtrl', ['$scope', function($scope){
  $scope.someData = [12, 15, 36, 8, 14, 20, 45, 18, 20, 27, 37, 48, 60, 30, 55, 80];
}]);

// View
<div bar-chart ng-model="someData"></div>
```


HISTORY.md
0.2.1 / 2015-01-29
==================

  * Fixing animate creating another rect

0.2.0 / 2015-01-29
==================

  * Ensuring non numeric values are evaluated as 0 instead of 100
  * Adding support for animations

0.1.0 / 2015-01-29
==================

  * Adding support for collections via attribute to indicate the key in which the value is.

0.0.1 / 2015-01-21
==================

  * Initial release


barchart.js

var scale = function(max, min, num) {
  return (100 * (num - min) / (max - min)) || 0;
}

var calcVertPosition = function(height, totalHeight) {
  return (totalHeight - height);
}

angular.module('ftBarChart', []).directive('barChart', function() {
  return {
    restrict: 'A',
    scope: {
      ngModel: '='
    },
    link: function(scope, elem, attrs) {

      var chartHeight = (attrs.chartHeight)? attrs.chartHeight : 90;
      var chartWidth = (attrs.chartWidth)? attrs.chartWidth : 400;
      var chartColor = (attrs.chartColor)? attrs.chartColor : '#000';
      var data = scope.ngModel;
      var max = Math.max.apply(null, data);
      var min = Math.min.apply(null, data);

      var parts = data.map(function(num) {
        return scale(max, min, num);
      });

      var x;
      var y;
      var height;
      var width = 100 / parts.length;

      var svgElem = angular.element('<svg width="' + chartWidth + '" height="' + chartHeight + '"/>');

      for (var i = 0; i < parts.length; i++) {
        var rect = document.createElementNS('http://www.w3.org/2000/svg', 'rect');
        x = 100 * (i / parts.length);
        y = 100 - parts[i];
        height = parts[i] + 1;
        rect.setAttribute('title', data[i]);
        rect.setAttribute('x', x + '%');
        rect.setAttribute('y', y + '%');
        rect.setAttribute('width', width + '%');
        rect.setAttribute('height', height + '%');
        rect.setAttribute('fill', chartColor);
        svgElem.append(rect);
      }

      elem.append( svgElem );
    }
  }
});

lib/angular.barchart.js
angular.module('ftBarChart', []).directive('barChart', ['$window', '$document', function($window, $document) {
  return {
    restrict: 'A',
    scope: {
      ngModel: '='
    },
    link: function(scope, elem, attrs) {

      // Initialize functions in link scope
      var scale = function(max, min, num) {
        return (100 * (num - min) / (max - min)) || 0;
      }

      var calcVertPosition = function(height, totalHeight) {
        return (totalHeight - height);
      }

      var drawChart = function(chartValues) {
        var max = Math.max.apply(null, chartValues);
        var min = Math.min.apply(null, chartValues);

        elem.html('');

        var parts = [];

        for ( var i = 0; i < chartValues.length; i++ ) {
          parts.push( scale(max, min, chartValues[i]) );
        }

        console.log(parts);

        var x;
        var y;
        var height;
        var width = 100 / parts.length;

        var svgElem = angular.element('<svg width="' + chartWidth + '" height="' + chartHeight + '"/>');

        for (var i = 0; i < parts.length; i++) {
          var rect = $document[0].createElementNS('http://www.w3.org/2000/svg', 'rect');
          x = 100 * (i / parts.length);
          y = 100 - parts[i];
          height = parts[i] + 1;
          rect.setAttribute('title', chartValues[i]);
          rect.setAttribute('x', x + '%');
          rect.setAttribute('y', y + '%');
          rect.setAttribute('width', width + '%');
          rect.setAttribute('height', height + '%');
          rect.setAttribute('fill', chartColor);
          svgElem.append(rect);
        }

        elem.append( svgElem );
      }

      var chartHeight = (attrs.chartHeight)? attrs.chartHeight : 90;
      var chartWidth = (attrs.chartWidth)? attrs.chartWidth : 400;
      var chartColor = (attrs.chartColor)? attrs.chartColor : '#000';
      var data = scope.ngModel;

      scope.$watch('ngModel', function(newVal, oldVal) {
        console.log('Data Change!');
        drawChart( newVal );
      });
    }
  }
}]);

lib/angular-barchart.js
angular.module('ftBarChart', [])
  .directive('barChart', function($window, $document) {
  return {
    restrict: 'A',
    scope: {
      ngModel: '=',
      objectKey: '@?barchartValueKey',
      shouldAnimate: '&?barchartAnimate',
      animateDuration: '&?barchartAnimateDuration'
    },
    link: function(scope, elem, attrs) {
      var dictMode = false;
      var shouldAnimate = !!scope.shouldAnimate();
      var animateDuration = '500ms';

      if (shouldAnimate && angular.isDefined(attrs.barchartAnimateDuration)) {
        animateDuration = scope.animateDuration();
      }

      if (angular.isObject(scope.ngModel[0])) {
        if (!angular.isDefined(attrs.barchartValueKey)) {
          throw new Error('You must define a barchart-value-key to extract the value from the dict!');
        }

        dictMode = true;
      }

      // Initialize functions in link scope
      var scale = function(max, min, num) {
        var result;

        if (angular.isNumber(num)) {
          result = (100 * (num - min) / (max - min)) || 0;
        } else {
          result = 0;
        }

        return result;
      };

      var drawChart = function(chartValues) {
        var max = Math.max.apply(null, chartValues);
        var min = Math.min.apply(null, chartValues);
        var i = 0;
        var j = 0;
        var chartLength = chartValues.length;
        var partsLength;

        elem.html('');

        var parts = [];

        for (; i < chartLength; i++) {
          parts.push(scale(max, min, chartValues[i]));
        }

        var x, y, height;
        var width = 100 / parts.length;
        var svgElem = angular.element('<svg width="' + chartWidth + '" height="' + chartHeight + '"/>');
        partsLength = parts.length;

        for (; j < partsLength; j++) {
          var rect = $document[0].createElementNS('http://www.w3.org/2000/svg', 'rect');
          x = 100 * (j / parts.length);
          y = 100 - parts[j];
          height = parts[j] + 1;
          rect.setAttribute('title', chartValues[j]);
          rect.setAttribute('x', x + '%');
          rect.setAttribute('y', y + '%');
          rect.setAttribute('width', width + '%');
          rect.setAttribute('height', height + '%');
          rect.setAttribute('fill', chartColor);

          if (shouldAnimate) {
            var animate = $document[0].createElementNS('http://www.w3.org/2000/svg', 'animate');

            animate.setAttribute('attributeName', 'y');
            animate.setAttribute('from', '100%');
            animate.setAttribute('to', y + '%');
            animate.setAttribute('dur', animateDuration);
            rect.appendChild(animate);
          }

          svgElem.append(rect);
        }

        elem.append(svgElem);
      };

      var getArray = function(newVal) {
        var array = newVal;

        if (dictMode) {
          array = newVal.map(function(obj) {
            return obj[scope.objectKey];
          });
        }

        return array;
      };

      var chartHeight = (attrs.chartHeight) ? attrs.chartHeight : 90;
      var chartWidth = (attrs.chartWidth) ? attrs.chartWidth : 400;
      var chartColor = (attrs.chartColor) ? attrs.chartColor : '#000';

      scope.$watch('ngModel', function(newVal) {
        drawChart(getArray(newVal));
      }, dictMode);
    }
  };
});
